nova_vm/src/ecmascript/builtins/data_view/data.rs

```diff
@@ -3,21 +3,151 @@
+                    Self::heap()
+                } else {
+                    Self::value(value as u32)
+                }
+            }
+            None => Self::auto(),
+        }
+    }
+}
+
+impl HeapMarkAndSweep for DataViewByteLength {
+    fn mark_values(&self, _queues: &mut WorkQueues) {
+        if *self == Self::heap() {
+            todo!()
+        }
+    }
+
+    fn sweep_values(&mut self, _compactions: &CompactionLists) {
+        if *self == Self::heap() {
+            todo!()
+        }
+    }
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) struct DataViewByteOffset(pub u32);
+
+impl DataViewByteOffset {
+    pub fn value(value: u32) -> Self {
+        Self(value)
+    }
+
+    /// A sentinel value of `u32::MAX` means that the byte offset is stored in
+    /// an associated map in the heap. This will most likely be a very rare
+    /// case, only applicable for 4GB+ buffers.
+    pub fn heap() -> Self {
+        Self(u32::MAX)
+    }
+}
+
+impl Default for DataViewByteOffset {
+    fn default() -> Self {
+        Self::value(0)
+    }
+}
+
+impl From<usize> for DataViewByteOffset {
+    fn from(value: usize) -> Self {
+        if value >= Self::heap().0 as usize {
+            Self::heap()
+        } else {
+            Self::value(value as u32)
+        }
+    }
+}
+
+impl HeapMarkAndSweep for DataViewByteOffset {
+    fn mark_values(&self, _queues: &mut WorkQueues) {
+        if *self == Self::heap() {
+            todo!()
+        }
+    }
+
+    fn sweep_values(&mut self, _compactions: &CompactionLists) {
+        if *self == Self::heap() {
+            todo!()
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct DataViewHeapData {
     pub(crate) object_index: Option<OrdinaryObject>,
+    // TODO: Add a helper function for a u32::MAX value which signifies an a under-construction value:
+    // See https://github.com/trynova/nova/pull/447#discussion_r1806247107 for reference.
+    /// ### [\[\[ViewedArrayBuffer\]\]](https://tc39.es/ecma262/#sec-properties-of-dataview-instances)
+    pub(crate) viewed_array_buffer: ArrayBuffer,
+    /// ### [\[\[ByteLength\]\]](https://tc39.es/ecma262/#sec-properties-of-dataview-instances)
+    pub(crate) byte_length: DataViewByteLength,
+    /// ### [\[\[ByteOffset\]\]](https://tc39.es/ecma262/#sec-properties-of-dataview-instances)
+    pub(crate) byte_offset: DataViewByteOffset,
+}
+
+impl Default for DataViewHeapData {
+    fn default() -> Self {
+        Self {
+            object_index: None,
+            viewed_array_buffer: ArrayBuffer::_def(),
+            byte_length: DataViewByteLength::default(),
+            byte_offset: DataViewByteOffset::default(),
+        }
+    }
 }
 
 impl HeapMarkAndSweep for DataViewHeapData {
     fn mark_values(&self, queues: &mut WorkQueues) {
         self.object_index.mark_values(queues);
+        self.viewed_array_buffer.mark_values(queues);
+        self.byte_length.mark_values(queues);
```

Comment:
**note:** You can just ignore byte_length and byte_offset here.

I think I've already changed this in main as well, but doing `let Self { a, b: _, c, d: _ } = self` (note the lack of `..` and usage of `x: _`) is a great way to make sure we're not accidentally forgetting to mark and/or sweep a field.